### PR TITLE
PLUGIN-399 add ways to compute end date time

### PR DESCRIPTION
**Pull request by yaojiefeng**

### CODE REVIEW

File: src/main/java/io/cdap/plugin/scd2/SCD2Plugin.java [hunk 1]
```diff
@@ -181,6 +187,10 @@ public String getEndDateField() {
       return endDateField;
     }
 
+    public long getEndDateOffset() {
+      return endDateOffset == null ? 1L : endDateOffset;
```
Comment (chtyim) [inline src/main/java/io/cdap/plugin/scd2/SCD2Plugin.java]: Do we need to guard against someone pass in a negative number?

Comment (yaojiefeng) [reply src/main/java/io/cdap/plugin/scd2/SCD2Plugin.java]: I added a check in validation to guard against a negative number

Comment (ghost) [reply src/main/java/io/cdap/plugin/scd2/SCD2Plugin.java]: Is it OK to have endDateOffset=0?

File: docs/SCD2-sparkcompute.md [hunk 2]
```diff
@@ -17,6 +17,9 @@ int, long, float, double, bytes or string type.
 **End Date Field:** The name of the end date field. The sorted results are iterated to compute the value of this field based
 on the start date.
 
+**End Date Offset:** The offset to compute the end date. The end date will be computed as the next start date minus this offset.
+The unit is microseconds.
```
Comment (albertshau) [inline docs/SCD2-sparkcompute.md]: microseconds is pretty user unfriendly. To specify a value for 1 second, would have to put in 1000000. One day would be even more unfriendly, with 9 zeros.

Does the UI have a widget that can be used to make this easier? If not, we should consider using the format supported by the TimeParser in hydrator-common to allow things like '1d', and '1s'.

File: src/main/java/io/cdap/plugin/scd2/SCD2Plugin.java (hunk 1 above)
Comment (yaojiefeng) [reply src/main/java/io/cdap/plugin/scd2/SCD2Plugin.java]: Yea 0 is valid, it is ok to have a end time which is equal to next start time

File: docs/SCD2-sparkcompute.md (hunk 2 above)
Comment (yaojiefeng) [reply docs/SCD2-sparkcompute.md]: The time parser in hydrator common only supports up to milliseconds, it only supports time unit to seconds. It is not feasible for our cases. Also checked with UI, there is currently no widget for this

File: docs/SCD2-sparkcompute.md [hunk 3]
```diff
@@ -18,7 +18,9 @@ int, long, float, double, bytes or string type.
 on the start date.
 
 **End Date Offset:** The offset to compute the end date. The end date will be computed as the next start date minus this offset.
-The unit is microseconds.
+The format is expected to be a number followed by an 'us', 'ms', 's', 'm', 'h', or 'd' specifying the time unit,
+with 'us' for microseconds, 'ms' for milliseconds, 's' for seconds, 'm' for minutes, 'h' for hours, and 'd' for days.
+By default, the offset is 1us.
```
Comment (albertshau) [inline docs/SCD2-sparkcompute.md]: instead of mentioning the default here, it's better to just set a default in the widget spec

**Review by albertshau — APPROVED** [review]
one minor comment

**Pull request merged**